src/trait_/common_ord.rs

```diff
@@ -214,79 +231,89 @@ pub fn build_ord_signature(
-
-	// Build separate `match` arms to compare different variants to each
-	// other. The index for these variants is used to determine which
-	// `Ordering` to return.
-	for (index, variant) in variants.iter().enumerate() {
-		let pattern = &variant.self_pattern();
+	use std::{borrow::Cow, ops::Deref};
+
+	use proc_macro2::Span;
+	use syn::{parse_quote, Expr, ExprLit, LitInt};
+
+	let mut discriminants = Vec::<Cow<Expr>>::with_capacity(variants.len());
+	let mut last_expression: Option<(Option<usize>, usize)> = None;
+
+	for variant in variants {
+		let discriminant = if let Some(discriminant) = variant.discriminant {
+			last_expression = Some((Some(discriminants.len()), 0));
+			Cow::Borrowed(discriminant)
+		} else {
+			let discriminant = match &mut last_expression {
+				Some((Some(expr_index), counter)) => {
+					let expr = &discriminants[*expr_index];
+					*counter += 1;
+					let counter = LitInt::new(&counter.to_string(), Span::call_site());
+					parse_quote! { (#expr) + #counter }
+				}
+				Some((None, counter)) => {
+					*counter += 1;
 
-		// The first variant is always `Less` then everything.
-		if index == 0 {
-			different.push(quote! {
-				#pattern => #less,
-			})
-		}
-		// The last variant is always `Greater` then everything.
-		else if index == variants.len() - 1 {
-			different.push(quote! {
-				#pattern => #greater,
-			})
-		}
-		// Any variant between the first and last.
-		else {
-			// Collect all variants that are `Less`.
-			let cases = variants
-				.iter()
-				.enumerate()
-				.filter(|(index_other, _)| *index_other < index)
-				.map(|(_, variant_other)| variant_other.other_pattern_skip().clone())
-				.collect();
-
-			// Build one match arm pattern with all variants that are `Greater`.
-			let pattern_less = PatOr {
-				attrs: Vec::new(),
-				leading_vert: None,
-				cases,
+					ExprLit {
+						attrs: Vec::new(),
+						lit: LitInt::new(&counter.to_string(), Span::call_site()).into(),
+					}
+					.into()
+				}
+				None => {
+					last_expression = Some((None, 0));
+					ExprLit {
+						attrs: Vec::new(),
+						lit: LitInt::new("0", Span::call_site()).into(),
+					}
+					.into()
+				}
 			};
 
-			// All other variants are `Less`.
-			different.push(quote! {
-				#pattern => match __other {
-					#pattern_less => #greater,
-					_ => #less,
-				},
-			});
-		}
+			Cow::Owned(discriminant)
+		};
+
+		discriminants.push(discriminant);
 	}
 
-	let rest = if incomparable.is_empty() {
-		quote!()
-	} else {
-		quote!(_ => unreachable!("incomparable variants should have already returned"),)
-	};
+	let variants = variants
+		.iter()
+		.zip(discriminants)
+		.map(|(variant, discriminant)| {
+			let pattern = variant.self_pattern();
+			let discriminant = discriminant.deref();
 
+			quote! {
+				#pattern => #discriminant
+			}
+		});
+
+	// `isize` is currently used by Rust as the default representation when none is
+	// defined.
+	let repr = repr.unwrap_or(Representation::ISize).to_token();
```

Comment:
This assumption is the only thing I'm not 100% sure is sound

Reply:
Hm, I have to think about this again, I've found an error in my logic ... again.

Reply:
Yeah so the issue is if have an enum with only empty variants, `repr(C)` and custom discriminents, and there is a platform that doesn't use `isize`, we might have an issue.

But I actually have a good answer to that: it's not a problem! Compilation will just fail if you use any funny expressions or literals that are bigger then `isize`. So there is no undefined behavior, it's just that we don't support `repr(C)` unless it has `isize`, which is currently the case on all platforms anyway.

Reply:
`as` only fails on debug, it wraps on release right?

Reply:
I'm loosing my mind a bit because I can't figure out the logic of what I wrote above there.
Still thinking ...

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

assuming, the `repr(u64)` is a `repr(C)` which for some reason behaves like a `repr(u64)`

Reply:
We built in some validation for `repr(C)` to make sure it's `isize`.